### PR TITLE
feat: `new URL(..., import.meta.url)` support

### PR DESCRIPTION
**Pull request by hyf0**

### CODE REVIEW

**Comment by hyf0** [comment]
* **#2710** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2710?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/src/ast_scanner/new_url.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    let Some(first_arg_string_literal) = expr
+      .arguments
+      .first()
+      .and_then(|arg| arg.as_expression().and_then(|expr| expr.as_string_literal()))
+    else {
+      return;
+    };
+    let path = &first_arg_string_literal.value;
+    let idx =
+      self.add_import_record(path, ImportKind::Import, expr.span, ImportRecordMeta::empty());
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/new_url.rs:37]: You may need to make sure the span of `new URL()` is unique now.

File: crates/rolldown/src/ast_scanner/new_url.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+impl<'me, 'ast: 'me> AstScanner<'me, 'ast> {
+  /// Handle exact `new URL('path', import.meta.url)` pattern
+  pub fn handle_new_url_with_string_literal_and_import_meta_url(
+    &mut self,
+    expr: &NewExpression<'ast>,
+  ) {
+    let is_callee_global_url = matches!(expr.callee.as_identifier(), Some(ident) if ident.name == "URL" && self.is_global_identifier_reference(ident));
+
+    let is_second_arg_import_meta_url = expr
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/new_url.rs:19]: I think you should early return if the conditions did not match

File: crates/rolldown/src/module_finalizers/scope_hoisting/mod.rs [hunk 3]
```diff
@@ -383,4 +384,53 @@ impl<'me, 'ast> ScopeHoistingFinalizer<'me, 'ast> {
...
+
+  /// Check if it is exact `new URL('path', import.meta.url)` pattern
+  fn is_new_url_with_string_literal_and_import_meta_url(
+    &self,
+    expr: &ast::Expression<'ast>,
+  ) -> bool {
+    let ast::Expression::NewExpression(expr) = expr else {
+      return false;
+    };
+    let is_callee_global_url = matches!(expr.callee.as_identifier(), Some(ident) if ident.name == "URL" && self.is_global_identifier_reference(ident));
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/module_finalizers/scope_hoisting/mod.rs:396]: Same here if the condition is not match early return it.

**Comment by magic-akari** [comment]
TODO:
- import.meta.resolve
- [proposal-asset-references](https://github.com/tc39/proposal-asset-references)

**Comment by IWANABETHATGUY** [comment]
> TODO:
> 
> * import.meta.resolve
> * [proposal-asset-references](https://github.com/tc39/proposal-asset-references)

Thank you for pointing that out.

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**